### PR TITLE
[rllib, tune] TrainingResult -> Dict, Removes C408 from flake8

### PR DESCRIPTION
**Pull request by richardliaw**
## What does this PR do?
This PR moves Tune's TrainingResult to a `dict` format, clarifying the confusion from previous users to report arbitrary fields.

Helper functions, like pretty printing, can be added into this class.

Tune is  changed to rely on dict syntax, with perhaps a few mandatory/auto-reported fields that  would be reasonable defaults for optimization mechanisms.

This would technically allow Tune users to report arbitrary values in a schemaless fashion, hence fracturing the csv support; but that should be OK. 

## TODOs: 
 - [x] Waiting for #2452 to merge.
 - [x] Move RLlib onto RLTrainingResult
    - [x] Fix `timesteps_total`/tests
 - [x] ~Change the connector from RLlib to Tune to be a dict~ No need to do this
 - [x] Move Tune onto dict/getter setter syntax
 - [ ] Fix documentation
    - [x] Remove neg_mean_loss and `timesteps_total` dependency

## Changes:
- Removed `neg_mean_loss` adapter

### CODE REVIEW

File: python/ray/rllib/agents/agent.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 from ray.rllib.evaluation.policy_evaluator import PolicyEvaluator
 from ray.rllib.utils import deep_update
 from ray.tune.registry import ENV_CREATOR, _global_registry
-from ray.tune.result import TrainingResult
+from ray.rllib.utils import RLTrainingResult as TrainingResult
```
Comment (ericl) [inline python/ray/rllib/agents/agent.py]: Let's rename it to rllib.evaluation.TrainingResult

File: python/ray/rllib/utils/result.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            in multi-agent RL.
+        mean_accuracy (float): (Optional) The current training accuracy
+            if applicable.
+        mean_validation_accuracy (float): (Optional) The current
+            validation accuracy if applicable.
+        mean_loss (float): (Optional) The current training loss if applicable.
+        neg_mean_loss (float): (Auto-filled) The negated current training loss.
+    """
+
+    @property
```
Comment (ericl) [inline python/ray/rllib/utils/result.py]: This is somewhat overkill, you can just do
```
def __getattr__(self, key):
    return self[key]
```

and raise the right ValueError otherwise.

File: python/ray/tune/result.py [hunk 3]
```diff
@@ -2,101 +2,120 @@
...
-        # (Auto-filled) The node ip of the machine hosting the
-        # training process.
-        "node_ip",
-
-        # (Auto=filled) The current hyperparameter configuration.
-        "config",
-    ])
 
-TrainingResult.__new__.__defaults__ = (None, ) * len(TrainingResult._fields)
+class TrainingResult(MutableMapping):
```
Comment (ericl) [inline python/ray/tune/result.py]: Can't this just extend a dict?

File: python/ray/tune/result.py [hunk 4]
```diff
@@ -2,101 +2,120 @@
...
+    def __repr__(self):
+        return "TrainingResult({})".format(self._store)
+
+    def as_dict(self):
+        """Retrieves copy ofinternal dict, used for JSON dumping."""
+        return copy.deepcopy(self._store)
+
+    @property
+    def done(self):
+        return self.get("done")
```
Comment (ericl) [inline python/ray/tune/result.py]: You can replace all of these with a `__getattr__`

File: python/ray/tune/result.py [hunk 5]
```diff
@@ -2,101 +2,120 @@
...
-    ])
 
-TrainingResult.__new__.__defaults__ = (None, ) * len(TrainingResult._fields)
+class TrainingResult(MutableMapping):
+    """Dict for storing results for Tune, with extra properties.
+
+    When using Tune with custom training scripts, you must periodically report
+    training status back to Ray by calling reporter(result).
+
+    Properties:
```
Comment (ericl) [inline python/ray/tune/result.py]: Attributes

File: python/ray/tune/trainable.py [hunk 6]
```diff
@@ -116,34 +112,22 @@ def train(self):
...
 
         now = datetime.today()
-        result = result._replace(
+        result = result.copy()
+        result.update(
             experiment_id=self._experiment_id,
             date=now.strftime("%Y-%m-%d_%H-%M-%S"),
             timestamp=int(time.mktime(now.timetuple())),
             training_iteration=self._iteration,
-            timesteps_total=self._timesteps_total,
```
Comment (ericl) [inline python/ray/tune/trainable.py]: Let's make sure timesteps_total is still calculated correctly, but it can be optional.

File: python/ray/tune/trainable.py [hunk 7]
```diff
@@ -217,8 +199,7 @@ def restore(self, checkpoint_path):
         metadata = pickle.load(open(checkpoint_path + ".tune_metadata", "rb"))
         self._experiment_id = metadata[0]
         self._iteration = metadata[1]
-        self._timesteps_total = metadata[2]
-        self._time_total = metadata[3]
+        self._time_total = metadata[2]
```
Comment (ericl) [inline python/ray/tune/trainable.py]: Revert this change.

File: python/ray/tune/trial.py [hunk 8]
```diff
@@ -271,21 +266,12 @@ def location_string(hostname, pid):
...
-        if self.last_result.mean_loss is not None:
-            pieces.append('{} loss'.format(
-                format(self.last_result.mean_loss, '.3g')))
-
-        if self.last_result.mean_accuracy is not None:
-            pieces.append('{} acc'.format(
-                format(self.last_result.mean_accuracy, '.3g')))
+        if self.last_result.metric is not None:
+            pieces.append('{}'.format(
+                format(self.last_result.metric, '.3g')))
```
Comment (ericl) [inline python/ray/tune/trial.py]: Please make sure this preserves the previous functionality.

**Review by ericl — CHANGES_REQUESTED** [review]
Overall: please make sure functionality does not regress here. Otherwise, it looks fine, but I would replace it with a dict instead of trying to define all the properties yourself.

File: python/ray/tune/result.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/tune/result.py]: It could, but you'd have to fix the Ray serialization because it would result to serializing by `__dict__`, which for a `dict(a=1)` object is empty

File: python/ray/tune/result.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/tune/result.py]: So the intent here was not to generalize the namedtuple functionality.

I think there are a small number of fields that should be special-cased, specifically the fields set and used by other parts of Tune. Everything else that the user touches should be only accessible via getitem/setitem syntax. The reasoning for this is that as a user, intuitively you would expect reading custom attributes and writing custom attributes to have the same semantics.

Is the `__getattr__` after subclassing dict standard?

Comment (richardliaw) [reply python/ray/tune/result.py]: So presumably if you replace `__getattr__`, you'll also want to set `__setattr__` to raise an exception if called?

File: python/ray/tune/result.py [hunk 9]
```diff
@@ -2,101 +2,17 @@
...
-When using ray.tune with custom training scripts, you must periodically report
-training status back to Ray by calling reporter(result).
 
-Most of the fields are optional, the only required one is timesteps_total.
-
-In RLlib, the supplied algorithms fill in TrainingResult for you.
-"""
-
-# Where ray.tune writes result files by default
+DONE = "done"
```
Comment (ericl) [inline python/ray/tune/result.py]: Move all the comments to these new enums?

**Review by ericl — APPROVED** [review]

**Comment by richardliaw** [comment]
jenkins retest this please

**Comment by richardliaw** [comment]
jenkins retest this please

**Comment by richardliaw** [comment]
Merging this - actor test and stress test failures are unrelated.
